src/errorpage.cc

```diff
@@ -1248,25 +1317,131 @@ ErrorState::BuildContent()
 MemBuf *ErrorState::ConvertText(const char *text, bool allowRecursion)
 {
     MemBuf *content = new MemBuf;
+    content->init();
+    convertAndWriteTo(text, *content, false, allowRecursion);
+    content->terminate();
+    assert((size_t)content->contentSize() == strlen(content->content()));
+    return content;
+}
+
+const char *
+ErrorState::NextCode(const char *p)
+{
+    assert(p);
+    do {
+        while (*p && *p != '%' && *p != '@') ++p;
+        if (*p == '@' && LogFormatStart.cmp(p, LogFormatStart.length()) != 0) {
+            ++p;
+            continue;
+        }
+        return p;
+    } while (1);
+}
+
+void
+ErrorState::convertAndWriteTo(const char *text, MemBuf &result, bool building_deny_info_url, bool allowRecursion)
+{
     const char *p;
     const char *m = text;
     assert(m);
-    content->init();
 
-    while ((p = strchr(m, '%'))) {
-        content->append(m, p - m);  /* copy */
-        const char *t = Convert(*++p, false, allowRecursion);   /* convert */
-        content->appendf("%s", t);   /* copy */
-        m = p + 1;          /* advance */
+    while (*(p = NextCode(m))) {
+        result.append(m, p - m);
+        if (*p == '%') {
+            const char *t = convert(++p, building_deny_info_url, allowRecursion);
+            result.appendf("%s", t);
+            m = p + 1;
+        } else {
+            assert(*p == '@');
+            handleLogFormat(p, result);
+            m = p;
+        }
     }
 
     if (*m)
-        content->appendf("%s", m);   /* copy tail */
+        result.appendf("%s", m);   /* copy tail */
+}
 
-    content->terminate();
+void
+ErrorState::handleLogFormat(const char *&start, MemBuf &result)
+{
+    assert(LogFormatStart.cmp(start, LogFormatStart.length()) == 0);
+
+    const char *p = start + LogFormatStart.length();
+    if (int tokenBytes = Format::Format::AssembleToken(p, result, al)) {
+        p += tokenBytes;
+        if (*p == '}') {
+            start = p + 1;
+            return;
+        }
 
-    assert((size_t)content->contentSize() == strlen(content->content()));
+        if (errorHandler_)
+            errorHandler_->handleError(SBuf("unterminated @Squid{"));
+    } else if (errorHandler_)
+        errorHandler_->handleError(ToSBuf("Cannot parse @Squid{logformat} in error page template near '", SBuf(start).substr(0, 100), "'"));
 
-    return content;
+    // Do not try to recover, stop interpreting the template
+    while (*p) ++p;
```

Comment:
However, this is another kind of 'recovery': it ignores all symbols up to the end of the input (the entire template file). I think that handleLogFormat() should not do that because it is a low-level and narrow-purpose method. it should stop parsing at this point and report an error (via throw an exception) instead: the exception handler would ignore the whole template then.

Reply:
Why? What are we going to gain?
Current code will provide a semi-interpreted template to the user. From the point of error and after the formatting codes will not interpreted. 
Is n't it better than providing to the end user an error page showing "Internal Error: Missing Template" or something similar? Will be more confusing for system admins.

Reply:
I think you a right: providing such half-parsed template is not perfect, but sending that "missing template" page instead would not be good either. Probably there should be a specific template for such kind of parsing errors (and this template should itself be quite simple to be checked at the boot time). However, adding such a template is a separate task.